### PR TITLE
Feat/parse multiple hcl files

### PR DESCRIPTION
**Pull request by teodora-sandu**
When trying to integrate the previous Golang parser with `gopherjs` in the snyk CLI we ran into some problems when returning the `VariableMap`, which has a complex type `(map[string]cty.Value)`. The code was getting more and more complicated and we realised it'd be easier to send a `map[string]interface{}` of file names and file contents to the parser. So now the parser:
- retrieves all files in a directory, corresponding to the files in a module (which will help when we implement modules)
- iterates through them once to get all variables in an array of `VariableMap`s
- merges the variables together into one `VariableMap`
- iterates through the files again to parse them using the merged `VariableMap`
This will also prepare us for when we decide to send one big merged file of files representing a Terraform module.

Some things to note:
- we don't include in the response files that failed due to internal errors (we may want to change this, but we have a lot of customers confused by irrelevant errors in their terminals)
- we only include user failures (such as from invalid Terraform)
- if a file fails to have its variables extracted due to a user error (i.e. invalid Terraform), we don't even try to parse it (this probably doesn't save much time but it will stop us from generating multiple errors for the same file)

There are some tests added in this PR, for the merge of the `VariableMap`s and the overall functionality, with different failure types to cover edge cases. If there are some cases that I missed, please flag them up.

The snyk CLI PR is https://github.com/snyk/snyk/pull/2705/commits.

### CODE REVIEW

File: terraform/errors.go [hunk 1]
```diff
@@ -7,9 +7,19 @@ type CustomError struct {
 }
 
 func (err *CustomError) Error() string {
+	// TODO: include more details here with user information
 	return err.message
 }
 
+func isUserError(err error) bool {
```
Comment (teodora-sandu) [inline terraform/errors.go:14]: I'm using this instead of returning `*CustomError` from the functions and accessing the field because apparently Golang does not treat `nil` as `nil` when the return type is an interface: https://go.dev/doc/faq#nil_error

File: terraform/errors.go [hunk 2]
```diff
@@ -7,9 +7,19 @@ type CustomError struct {
 }
 
 func (err *CustomError) Error() string {
+	// TODO: include more details here with user information
```
Comment (teodora-sandu) [inline terraform/errors.go:10]: Leaving this here in case we want to enhance the errors in future

File: terraform/hcl2json.go [hunk 3]
```diff
@@ -6,9 +6,55 @@ import (
...
+		}
+		variablesMaps = append(variablesMaps, variableMap)
+	}
+
+	// merge variables so they can be used across multiple files
+	variableMap := mergeVariables(variablesMaps)
+
+	for fileName, fileContent := range files {
+		if failedFiles[fileName] == nil {
+			parsedJson, err := parseHclToJson(fileName, fileContent.(string), variableMap)
```
Comment (ipapast) [inline terraform/hcl2json.go:38]: noob q: isn't fileContent already a string here? or we are just casting it "just in case" again, to be safe?

File: terraform/hcl2json_test.go [hunk 4]
```diff
@@ -561,20 +561,171 @@ variable "test" {
...
+			},
+			extractErr: &CustomError{
+				message:   "Internal error",
+				errors:    []error{},
+				userError: false,
+			},
+			expected: map[string]interface{}{
+				"failedFiles": map[string]interface{}{},
+				"parsedFiles": map[string]interface{}{
+					"fail.tf":  jsonOutput, // it's intentional for files that fail with internal errors at extraction time to still try to parse
```
Comment (ipapast) [inline terraform/hcl2json_test.go:656]: correct  me if I'm wrong:
user errors are errors that happened because of bad hcl, for example, but internal errors can be anything else that happened during parsing. is that right?
I'm curious  about this comment, what does it mean? if a file failed with an internal error when it was under the extractVariables method, we will still try to parse it? don't we skip it when we are checking `if failedFiles[fileName] == nil`?

Comment (teodora-sandu) [reply terraform/hcl2json_test.go:656]: So at the moment `failedFiles` contains errors we want to show the user - internal errors don't get shown to users at all. I'm interested to see if you think this is a good idea or not btw

When we iterate through the files to extract variables, we check if there's an error and if there is one:
- and it's a user error, we include it in `failedFiles` (i.e. `failedFiles[fileName] != nil`) and don't push anything to the array of `VariableMap`s
- and it's an internal error, we don't include it in `failedFiles` (see first paragraph in this comment) and don't push anything to the array of `VariableMap`s
Then, we iterate again through the files to parse them (this step), but only if there weren't any user errors in the previous step to extract variables (i.e. `failedFiles[fileName] == nil`). So we still try to parse them if there was an internal error in case there's a tiny chance the error was a flake.

Does this make sense? I can add a comment to the `ParseModule` function to explain this a bit, but will wait to see what you think

Comment (ipapast) [reply terraform/hcl2json_test.go:656]: yes, thanks, I think it hadn't clicked to that we can have "flakes". have you seen examples of this before I assume?

for a comment, even if you update this line 
"// it's intentional for files that fail with internal errors at extraction time to still try to parse" with something extra is still good e.g.
 "// it's intentional for files that fail with internal errors at extraction time to still try to parse as the internal error can be a flake"

**Review by ipapast — APPROVED** [review]

File: terraform/hcl2json_test.go (hunk 4 above)
Comment (teodora-sandu) [reply terraform/hcl2json_test.go:656]: I was thinking about some of the feedback we'd been getting during usability testing - see https://snyk.slack.com/archives/C02JMMTLUF9/p1643647684197409. At the moment we print out a lot of errors that the user can't really action, so this separation was meant to help with that.
